### ISSUE TITLE
feat(finetune): add CPT — Continual Pre-Training on raw text corpora

### ISSUE THREAD

**Issue by noahgift** [paiml/aprender#448]
## Summary

Add Continual Pre-Training (CPT) as a fine-tuning method. CPT is the first
stage of the 4-layer adaptation pipeline (CPT → SFT → DPO → RLVR) and
teaches domain knowledge from raw text without instruction formatting.

## CLI

```bash
apr finetune model.apr --method cpt \
  --data domain-corpus.txt \
  --epochs 1 \
  --learning-rate 2e-5 \
  -o cpt-model.apr
```

## Data Format

Plain text, one document per line (no instruction/response pairs):
```
Document text here about legal case law...
Another document about medical procedures...
```

## Key Differences from SFT

- Causal LM objective (next-token prediction) on raw text
- No chat template applied
- Typically lower learning rate (1e-5 to 5e-5)
- Longer training on larger corpora
- LoRA/QLoRA supported for efficient domain adaptation

## Use Cases

- Legal domain adaptation (case law, statutes)
- Medical domain (clinical notes, research papers)
- Code domain (proprietary codebase pre-training)
- Financial domain (earnings calls, SEC filings)

## Implementation

1. Add `cpt` variant to FinetuneMethod enum
2. Data loader for plain text (no JSON parsing, just line-by-line)
3. Causal LM training loop (same as SFT but without template)
4. Support LoRA and QLoRA for CPT
5. Checkpoint saving compatible with subsequent SFT stage

## Specification

See `docs/specifications/components/finetune.md` §5.1.

### Contract: `continual-pretraining-v1.yaml`

Extends existing `classification-finetune-v1.yaml` pattern for the CPT training loop.

```yaml
metadata:
  description: "Continual Pre-Training — domain knowledge acquisition from raw text"
  references:
    - "Gururangan et al. (2020) Don't Stop Pretraining"
  depends_on:
    - "cross-entropy-kernel-v1"
    - "adamw-kernel-v1"
    - "lora-algebra-v1"

equations:
  causal_lm_loss:
    formula: "L = -1/T * Σ_{t=1}^{T} log P(x_t | x_{<t})"
    invariants:
      - "L >= 0 (cross-entropy is non-negative)"
      - "L decreases over training epochs (on training set)"
      - "T = sequence_length"

  learning_rate_schedule:
    formula: "η_t = η_0 * warmup(t) * cosine_decay(t)"
    invariants:
      - "η_t > 0 for all t"
      - "η_t <= η_0 (never exceeds initial LR)"

proof_obligations:
  - type: bound
    property: "Loss non-negativity"
    formal: "causal_lm_loss(logits, targets) >= 0"
  - type: monotonicity
    property: "Training loss decrease"
    formal: "E[L(epoch_t)] < E[L(epoch_{t-1})] for sufficient data"
  - type: invariant
    property: "No chat template applied"
    formal: "input_text == raw_text (no template wrapping)"
  - type: invariant
    property: "Checkpoint compatibility"
    formal: "CPT checkpoint loadable by SFT stage"

falsification_tests:
  - id: FALSIFY-CPT-001
    rule: "Loss non-negative"
    prediction: "causal_lm_loss >= 0 for all inputs"
    if_fails: "Cross-entropy computation bug"
  - id: FALSIFY-CPT-002
    rule: "No template contamination"
    prediction: "raw text passed to model without ChatML/instruction wrapping"
    if_fails: "Template code path leaking into CPT"
  - id: FALSIFY-CPT-003
    rule: "Checkpoint roundtrip"
    prediction: "save(CPT checkpoint) → load(for SFT) succeeds"
    if_fails: "Checkpoint format incompatibility between stages"
```

### Binding Requirements

Add to `contracts/entrenar/binding.yaml`:
```yaml
  - contract: continual-pretraining-v1.yaml
    equation: causal_lm_loss
    module_path: "entrenar::finetune::cpt"
    function: cpt_training_loop
    status: not_implemented
```

### Verification Targets
- L2: 3 falsification tests
- L3: Kani harness for loss non-negativity (reuse cross-entropy-kernel harness)
